src/main/java/seedu/address/model/ModelManager.java

```diff
@@ -128,21 +144,166 @@ public boolean hasEntry(Entry entry) {
         return entryBook.hasPerson(entry);
     }
 
+    @Override
+    public boolean hasEntryEqualTo(Entry entry) {
+        requireNonNull(entry);
+        return entryBook.hasEntryEqualTo(entry);
+    }
+
     @Override
     public void deleteEntry(Entry target) {
         entryBook.removePerson(target);
     }
 
     @Override
-    public void addEntry(Entry entry) {
-        entryBook.addPerson(entry);
+    public Entry addEntry(Entry entry) {
+
+        /**
+         * Represents a builder that tries out replacement candidates and accepts suitable ones
+         */
+        abstract class Candidate<T> {
+            private T value;
+
+            Candidate(T initialValue) {
+                this.value = initialValue;
+            }
+
+            abstract Candidate<T> tryout(String candidate);
+
+            void setValue(T value) {
+                this.value = value;
+            }
+
+            T get() {
+                return this.value;
+            }
+        }
+
+        /**
+         * Represents a title that tries out replacement candidates and accepts suitable ones
+         */
+        class CandidateTitle extends Candidate<Title> {
+
+            private CandidateTitle() {
+                super(new Title("Untitled"));
+            }
+
+            @Override
+            Candidate<Title> tryout(String candidate) {
+                try {
+                    setValue(ParserUtil.parseTitle(Optional.of(candidate)));
+                } catch (ParseException pe) {
+                    logger.warning("Failed to replace title candidate. " + pe.getMessage());
+                }
+                return this;
+            }
+
+        }
+
+        /**
+         * Represents a description that tries out replacement candidates and accepts suitable ones
+         */
+        class CandidateDescription extends Candidate<Description> {
+
+            private CandidateDescription() {
+                super(new Description("No description"));
+            }
+
+            @Override
+            Candidate<Description> tryout(String candidate) {
+                try {
+                    setValue(ParserUtil.parseDescription(Optional.of(candidate)));
+                } catch (ParseException pe) {
+                    logger.warning("Failed to replace description candidate. " + pe.getMessage());
+                }
+                return this;
+            }
+
+        }
+
+        // Initial data
+        final Title title = entry.getTitle();
+        final Description description = entry.getDescription();
+        final boolean noTitleOrDescription = title.isEmpty() || description.isEmpty();
+        final String urlString = entry.getLink().value;
+
+        // Candidates to replace empty title and description
+        Candidate<Title> candidateTitle = new CandidateTitle();
+        Candidate<Description> candidateDescription = new CandidateDescription();
+
+        // 3rd choice - extract candidates just from URL
```

Comment:
The intended meaning is that generally for both `Title` and `Description` the 3rd choice is to extract candidates from the URL, but within a general choice a `Title` or `Description` could have multiple choices, but I see how it can be confusing - I'll try to make comments more clear.

Reply:
Fixed in bf6b79d